src/compiler/crystal/compiler.cr

```diff
@@ -346,13 +350,37 @@ module Crystal
         object_arg = Process.quote_windows(object_names)
         output_arg = Process.quote_windows("/Fe#{output_filename}")
 
+        cl = CL
         link_args = [] of String
+
+        # if the compiler and the target both have the `msvc` flag, we are not
+        # cross-compiling and therefore we should attempt detecting MSVC's
+        # standard paths
+        {% if flag?(:msvc) %}
+          if msvc_path = Crystal::System::VisualStudio.find_latest_msvc_path
+            if win_sdk_libpath = Crystal::System::WindowsSDK.find_win10_sdk_libpath
+              host_bits = {{ flag?(:bits64) ? "Hostx64" : "Hostx86" }}
+              target_bits = program.has_flag?("bits64") ? "x64" : "x86"
+
+              # MSVC build tools and Windows SDK found; recreate `LIB` environment variable
+              # that is normally expected on the MSVC developer command prompt
+              link_args << Process.quote_windows("/LIBPATH:#{msvc_path}\\atlmfc\\lib\\#{target_bits}")
+              link_args << Process.quote_windows("/LIBPATH:#{msvc_path}\\lib\\#{target_bits}")
+              link_args << Process.quote_windows("/LIBPATH:#{win_sdk_libpath}\\ucrt\\#{target_bits}")
+              link_args << Process.quote_windows("/LIBPATH:#{win_sdk_libpath}\\um\\#{target_bits}")
+
+              # use exact path for compiler instead of relying on `PATH`
+              cl = Process.quote_windows("#{msvc_path}\\bin\\#{host_bits}\\#{target_bits}\\cl.exe")
+            end
+          end
+        {% end %}
+
         link_args << "/DEBUG:FULL /PDBALTPATH:%_PDB%" unless debug.none?
         link_args << lib_flags
         @link_flags.try { |flags| link_args << flags }
 
         args = %(/nologo #{object_arg} #{output_arg} /link #{link_args.join(' ')}).gsub("\n", " ")
-        cmd = "#{CL} #{args}"
+        cmd = "#{cl} #{args}"
 
```

Comment:
Below this there's a line `cmd = "#{CL}` - that's probably wrong to keep using the old path there